### PR TITLE
POL-101- Azure Inefficient Instance Utilization using Log Analytics

### PR DESCRIPTION
**Pull request by KTebrewal**
### Description

This Policy Template uses performance metrics from Log Analytics from the last 30 days to identify underutilized instances and provides rightsizing recommendations. Once recommendations are generated, instances can be rightsized in an automated manner or after approval. This is meant to be run as a weekly policy.

### Issues Resolved

Removed the tagging escalation/action
Added Resize action for underutilized instances
Changed the authentication to credential services

### Contribution Check List

- [x] All tests pass.
- [x] New functionality includes testing.
- [x] New functionality has been documented in the README if applicable
- [x] New functionality has been documented in CHANGELOG.MD

### Running template Link

https://governance.rightscale.com/org/1105/projects/60073/policy-incidents/5e158a1732401200016de440

### Screenshot

image

### CODE REVIEW

File: cost/azure/instances_log_analitics_utilization/README.md [hunk 1]
```diff
@@ -1,66 +1,52 @@
-## Azure Instances Utilization from Log Analytics
+# Azure Instances Utilization from Log Analytics
```
Comment (cdwilhelm) [inline cost/azure/instances_log_analitics_utilization/README.md]: the readme header should match the new policy name

File: cost/azure/instances_log_analitics_utilization/azure_instance_log_analytics_utilization.pt [hunk 2]
```diff
@@ -1,10 +1,16 @@
...
-long_description "Version: 1.1"
+short_description "Gathers instance utilization data from Azure Log Analytics from the last 30 days to identify underutilized instances and provides rightsizing recommendations. Once recommendations are generated, instances can be rightsized in an automated manner or after approval. See the [README](https://github.com/flexera/policy_templates/tree/master/cost/azure/instances_log_analitics_utilization) and [docs.rightscale.com/policies](https://docs.rightscale.com/policies/) to learn more."
+long_description ""
 severity "low"
 category "Cost"
+info(
+  version: "2.0",
+  provider: "Azure",
+  service: "Compute",
+  policy_set: ""
```
Comment (cdwilhelm) [inline cost/azure/instances_log_analitics_utilization/azure_instance_log_analytics_utilization.pt]: use this value for the policy set.

File: cost/azure/instances_log_analitics_utilization/azure_instance_log_analytics_utilization.pt [hunk 3]
```diff
@@ -1,10 +1,16 @@
-name "Azure Instances Utilization from Log Analytics"
+name "Azure Inefficient Instance Utilization using Log Analytics"
 rs_pt_ver 20180301
 type "policy"
-short_description "Gathers instance utilization data from Azure Log Analytics and tags underutilized instances.  See the [README](https://github.com/rightscale/policy_templates/tree/master/cost/azure/instances_log_analitics_utilization) and [docs.rightscale.com/policies](http://docs.rightscale.com/policies/) to learn more."
-long_description "Version: 1.1"
+short_description "Gathers instance utilization data from Azure Log Analytics from the last 30 days to identify underutilized instances and provides rightsizing recommendations. Once recommendations are generated, instances can be rightsized in an automated manner or after approval. See the [README](https://github.com/flexera/policy_templates/tree/master/cost/azure/instances_log_analitics_utilization) and [docs.rightscale.com/policies](https://docs.rightscale.com/policies/) to learn more."
```
Comment (cdwilhelm) [inline cost/azure/instances_log_analitics_utilization/azure_instance_log_analytics_utilization.pt]: Use this value for the short_description

Checks for instances that are idle for the last 30 days and terminates them after approval.

File: cost/azure/instances_log_analitics_utilization/azure_instance_log_analytics_utilization.pt [hunk 4]
```diff
@@ -73,28 +60,23 @@ end
...
-    client_secret cred("AZURE_APPLICATION_KEY")
-    additional_params do {
-      "resource" => "https://management.azure.com/"
-    } end
-  end
+#authenticate with Azure
+credentials "azure_auth" do
+  schemes "oauth2"
+  label "Azure"
+  description "Azure Resource Manager"
```
Comment (cdwilhelm) [inline cost/azure/instances_log_analitics_utilization/azure_instance_log_analytics_utilization.pt]: change description to **Select the Azure Resource Manager Credential from the list.**

File: cost/azure/instances_log_analitics_utilization/azure_instance_log_analytics_utilization.pt [hunk 5]
```diff
@@ -73,28 +60,23 @@ end
...
-    client_secret cred("AZURE_APPLICATION_KEY")
-    additional_params do {
-      "resource" => "https://api.loganalytics.io"
-    } end
-  end
+#authenticate with Log Analytics
+credentials "log_analytics_auth" do
+  schemes "oauth2"
+  label "Log Analytics"
+  description "Please select the credentials for Log Analytics"
```
Comment (cdwilhelm) [inline cost/azure/instances_log_analitics_utilization/azure_instance_log_analytics_utilization.pt]: Please change the description to **Select the Azure Resource Manager Credential for Log Analytics.**

File: cost/azure/instances_log_analitics_utilization/azure_instance_log_analytics_utilization.pt [hunk 6]
```diff
@@ -73,28 +60,23 @@ end
...
-    client_id cred("AZURE_APPLICATION_ID")
-    client_secret cred("AZURE_APPLICATION_KEY")
-    additional_params do {
-      "resource" => "https://api.loganalytics.io"
-    } end
-  end
+#authenticate with Log Analytics
+credentials "log_analytics_auth" do
+  schemes "oauth2"
+  label "Log Analytics"
```
Comment (cdwilhelm) [inline cost/azure/instances_log_analitics_utilization/azure_instance_log_analytics_utilization.pt]: use Azure Log Analytics

File: cost/azure/instances_log_analitics_utilization/azure_instance_log_analytics_utilization.pt [hunk 7]
```diff
@@ -318,87 +342,125 @@ script "js_oms_resources", type: "javascript" do
...
+  for ( i = 0; i < ds_merged_metrics.length; i++){
+    var instance = ds_merged_metrics[i]
+    var cpu_average = instance.cpu_average
+    var mem_average = instance.mem_average
+    var next_vm_size = ds_azure_instance_size_map[instance.vmSize].down
+    if(next_vm_size == null && typeof next_vm_size == "undefined" ){
+      next_vm_size="N/A";
+    }
+    if (cpu_average < param_avg_cpu || mem_average < param_avg_used_memory){
+      inefficient_utilization_results.push({
```
Comment (cdwilhelm) [inline cost/azure/instances_log_analitics_utilization/azure_instance_log_analytics_utilization.pt]: sort by the subscriptonName then location.

File: cost/azure/instances_log_analitics_utilization/azure_instance_log_analytics_utilization.pt (hunk 2 above)
Comment (KTebrewal) [reply cost/azure/instances_log_analitics_utilization/azure_instance_log_analytics_utilization.pt]: I have collected the value for policy set from the spreadsheet and updated that in the policy - Inefficient Instance Usage

File: cost/azure/instances_log_analitics_utilization/README.md (hunk 1 above)
Comment (KTebrewal) [reply cost/azure/instances_log_analitics_utilization/README.md]: I have updated the Readme header and also updated the root Readme file.

File: cost/azure/instances_log_analitics_utilization/azure_instance_log_analytics_utilization.pt (hunk 3 above)
Comment (KTebrewal) [reply cost/azure/instances_log_analitics_utilization/azure_instance_log_analytics_utilization.pt]: I am using the below short description that i got from the spreadsheet.
"This checks inefficient instance utilization using provided CPU and Memory thresholds.  Instances matching the criteria can be resized after user approval."

File: cost/azure/instances_log_analitics_utilization/azure_instance_log_analytics_utilization.pt (hunk 4 above)
Comment (KTebrewal) [reply cost/azure/instances_log_analitics_utilization/azure_instance_log_analytics_utilization.pt]: Updated this description.

File: cost/azure/instances_log_analitics_utilization/azure_instance_log_analytics_utilization.pt (hunk 5 above)
Comment (KTebrewal) [reply cost/azure/instances_log_analitics_utilization/azure_instance_log_analytics_utilization.pt]: Updated this description.

File: cost/azure/instances_log_analitics_utilization/azure_instance_log_analytics_utilization.pt (hunk 6 above)
Comment (KTebrewal) [reply cost/azure/instances_log_analitics_utilization/azure_instance_log_analytics_utilization.pt]: Updated this label.

File: cost/azure/instances_log_analitics_utilization/azure_instance_log_analytics_utilization.pt (hunk 7 above)
Comment (KTebrewal) [reply cost/azure/instances_log_analitics_utilization/azure_instance_log_analytics_utilization.pt]: Can you please give a little more clarity on this one? What should be the sorting based on?

File: cost/azure/instances_log_analitics_utilization/azure_instance_log_analytics_utilization.pt (hunk 3 above)
Comment (cdwilhelm) [reply cost/azure/instances_log_analitics_utilization/azure_instance_log_analytics_utilization.pt]: Yes, I understand.  All policies with the same policy set should have the same value.  This is new behavior for the catalog with we show policy sets.  Please use the value I provided.  thank you.

File: cost/azure/instances_log_analitics_utilization/azure_instance_log_analytics_utilization.pt (hunk 7 above)
Comment (cdwilhelm) [reply cost/azure/instances_log_analitics_utilization/azure_instance_log_analytics_utilization.pt]: I would like to see the sorting by Subscription Name and the location.  There could be multiple locations in the same subscription. Sort by subscription name first, then location.

Comment (KTebrewal) [reply cost/azure/instances_log_analitics_utilization/azure_instance_log_analytics_utilization.pt]: I have added the code for same. It is now sorting by subscription and location.

**Pull request merged**